racker/salus-telemetry-monitor-management
# Calculate and maintain zone binding counts in SQL

# Resolves

https://jira.rax.io/browse/SALUS-358

# What

Rather than use a brittle, DIY counter in etcd, we'll instead use etcd to simply retrieve a list of the active poller's per remote monitoring zone and use the `BoundMonitor` table as the single source of truth for bindings per envoy-poller.

The motivation to make this change was observing my local system having troubles binding a new monitor to the 1000 resources I had left over from envoy-ambassador stress testing. The `ZoneStorage` code was getting itself into a tight loop of operations like the following with numerous collisions trying to increment the binding count:

```
2020-08-28 13:22:10.821 DEBUG [salus-telemetry-monitor-management,,,] 78771 --- [         etcd22] c.r.s.t.etcd.services.ZoneStorage        : Putting new bound count=894 for resource=development:0 in zone=ResolvedZone(name=dev, tenantId=aaaaaa)
2020-08-28 13:22:10.821 DEBUG [salus-telemetry-monitor-management,,,] 78771 --- [         etcd27] c.r.s.t.etcd.services.ZoneStorage        : Changing bound count of resource=development:0 in zone=ResolvedZone(name=dev, tenantId=aaaaaa) by amount=-1
2020-08-28 13:22:10.827 DEBUG [salus-telemetry-monitor-management,,,] 78771 --- [         etcd30] c.r.s.t.etcd.services.ZoneStorage        : Re-trying incrementing bound count of resource=development:0 in zone=ResolvedZone(name=dev, tenantId=aaaaaa) due to collision
```

# How

[The first commit](https://github.com/racker/salus-telemetry-monitor-management/commit/965b6cace8a3923034b6ae8cd0b5087b74410d47) only a refactoring to consolidate onto a single method of `MonitorManagement` for locating a least-loaded envoy-poller.

The second, big commit includes a new `ZoneAllocationResolver` component, which is allocated and re-used before each iterative operation done in monitor management. For example, if a monitor is about to get bound to 10,000 tenant resources, then a single instance of `ZoneAllocationResolver` is utilized for that. Each subsequent operation gets a fresh instance of that component though.

As you'll see in the javadoc, the `ZoneAllocationResolver` uses a minimal, cheap query into etcd to zero out the known, active envoy-pollers. Then it uses the new named queries on `BoundMonitor` to retrieve an aggregate count of bound monitors per envoy-poller resourceid for a requested zone.

The solution should scale to handle large numbers of tenant resources. The etcd query is small and constant. The `BoundMonitor` monitor table becomes the single source of truth and the remaining "expensive" operation is done within SQL, optimized by new indices, and the snapshot of those counts is used in-memory for any size of monitor-resource-zone binding operation.

# How to test

Updated existing unit tests of remote monitor binding to confirm the `pollerResourceId` was set as expected.

Added a new unit class for the new `ZoneAllocationResolver`.

# Depends on

- https://github.com/racker/salus-telemetry-model/pull/151
- https://github.com/racker/salus-telemetry-etcd-adapter/pull/67
